iroh-rpc/src/client.rs

```diff
@@ -0,0 +1,331 @@
+
+        let (header, stream) = match receiver.await.expect("Sender not to be dropped.") {
+            SenderType::Stream { header, stream } => (header, stream),
+            SenderType::Error(e) => return Err(e),
+            _ => return Err(RpcError::UnexpectedResponseType),
+        };
+        // examine header here and determine if we want to accept file
+        Ok(InStream::new(header, stream, self.command_sender.clone()))
+    }
+
+    /// Get the peer id from the Client's address book, based on the namespace
+    pub fn get_peer_id(&self, namespace: &str) -> Result<PeerId, RpcError> {
+        match self.addresses.get(namespace) {
+            Some((_, id)) => Ok(*id),
+            None => Err(RpcError::NamespaceNotFound(namespace.into())),
+        }
+    }
+
+    /// Get the multiaddr from the Client's address book, based on the namespace
+    pub fn get_multiaddr(&self, namespace: &str) -> Result<Multiaddr, RpcError> {
+        match self.addresses.get(namespace) {
+            Some((addr, _)) => Ok(addr.clone()),
+            None => Err(RpcError::NamespaceNotFound(namespace.into())),
+        }
+    }
+}
+
+struct Iter {
+    num: u64,
+}
+
+impl Iter {
+    fn new() -> Self {
+        Iter { num: 0 }
+    }
+}
+
+impl Iterator for Iter {
+    type Item = u64;
+    fn next(&mut self) -> Option<Self::Item> {
+        self.num += 1;
+        Some(self.num)
+    }
+}
+
+/// Dial a single address
+async fn dial(
+    mut command_sender: mpsc::Sender<Command>,
+    addr: Multiaddr,
+    peer_id: PeerId,
+) -> Result<(), RpcError> {
+    let (sender, receiver) = oneshot::channel();
+    command_sender
+        .send(Command::Dial {
+            peer_id,
+            peer_addr: addr,
+            sender,
+        })
+        .await
+        .expect("Receiver to not be dropped.");
+    match receiver.await.expect("Sender to not be dropped.") {
+        SenderType::Ack => Ok(()),
+        SenderType::Error(e) => Err(e),
+        _ => Err(RpcError::UnexpectedResponseType),
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use libp2p::identity::Keypair;
+
+    use crate::config::RpcConfig;
+    use crate::handler;
+    use crate::rpc_from_config;
+    use crate::serde::{deserialize_request, serialize_response, Deserialize};
+    use crate::swarm;
+
+    #[derive(Deserialize, Serialize, Debug, Clone)]
+    struct GetRequest {
+        resource_id: String,
+    }
+
+    #[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
+    struct GetResponse {
+        response: String,
+    }
+
+    async fn get(
+        state: handler::State<String>,
+        _stream_id: Option<u64>,
+        param: Vec<u8>,
+    ) -> Result<Vec<u8>, RpcError> {
+        let _: GetRequest = deserialize_request(&param)?;
+
+        let bytes = serialize_response(GetResponse {
+            response: state.clone(),
+        })?;
+        Ok(bytes)
+    }
+
+    #[tokio::test]
+    async fn client_example() {
```

Comment:
would be nice to add the following

- one case where it uses tcp, so we know it can work between processes
- one case wehre it the other side sends a different message as well
- error case for the rpc method & for sending an unknown method